Use the correct paint methods for painting a section rect of RenderWidgets

Ahmad-S792 opened the pull request:
#### fe7889ca7da926d131e576beed57733828295b82
<pre>
Use the correct paint methods for painting a section rect of RenderWidgets

Use the correct paint methods for painting a section rect of RenderWidgets
<a href="https://bugs.webkit.org/show_bug.cgi?id=246939">https://bugs.webkit.org/show_bug.cgi?id=246939</a>

Reviewed by Simon Fraser.

Merge - <a href="https://src.chromium.org/viewvc/blink?revision=181068&amp">https://src.chromium.org/viewvc/blink?revision=181068&amp</a>;view=revision

Previously, it was using a paint invalidation method. That is less
efficient, and also wrong if the renderer is a squashed layer.

* Source/WebCore/rendering/RenderElement.h: Update &quot;paint&quot; function to to have &quot;paintOffset&quot; and also add comment for explanation
* Source/WebCore/rendering/RenderObject.h: Update Comment and remove &quot;selectionRect&quot; function
* Source/WebCore/rendering/RenderWidget.cpp: Fix paint logic for selected widgets and remove &quot;FIXME&quot; about selectionRect
* LayoutTests/editing/selection/iframe-select.html: Added Test Case
* LayoutTests/editing/selection/iframe-select-expected.html: Added Test Case Expectations

Canonical link: <a href="https://commits.webkit.org/256191@main">https://commits.webkit.org/256191@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/20088596f47c23a2bb019881f6ebc3c9dc85fc18

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95006 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4150 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27910 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104608 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164863 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98997 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4237 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/32329 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87309 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100511 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100673 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3072 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81522 "Built successfully") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30054 "Found 1 new test failure: media/encrypted-media/mock-MediaKeySystemAccess.html (failure)") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84987 "Found 1 new API test failure: TestWebKitAPI.UserContentWorld.IsolatedWorldPlugIn (failure)") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/84542 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72947 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38740 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/18366 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36564 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19647 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4279 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40496 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42475 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38881 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/editing/selection/iframe-select.html:
```diff
@@ -0,0 +1,7 @@
+<!DOCTYPE html>
+<iframe id="iframe" style="width: 100px; height: 100px; position: absolute; top: 50px; left: 50px" frameBorder=0 srcdoc="<div></div>"></iframe>
```
This test needs to add `will-change: transform` to the style to exercise the correct code path.

Ahmad-S792 replied:
Done.

smfr reviewed (changes requested):
I tested the change; with the suggested change to the test I can reproduce the bug, and the patch fixes it.

Ahmad-S792 committed:
Use the correct paint methods for painting a section rect of RenderWidgets

Use the correct paint methods for painting a section rect of RenderWidgets
https://bugs.webkit.org/show_bug.cgi?id=246939

Reviewed by Simon Fraser.

Merge - https://src.chromium.org/viewvc/blink?revision=181068&view=revision

Previously, it was using a paint invalidation method. That is less
efficient, and also wrong if the renderer is a squashed layer.

* Source/WebCore/rendering/RenderElement.h: Update "paint" function to to have "paintOffset" and also add comment for explanation
* Source/WebCore/rendering/RenderObject.h: Update Comment and remove "selectionRect" function
* Source/WebCore/rendering/RenderWidget.cpp: Fix paint logic for selected widgets and remove "FIXME" about selectionRect
* LayoutTests/editing/selection/iframe-select.html: Added Test Case
* LayoutTests/editing/selection/iframe-select-expected.html: Added Test Case Expectations

Canonical link: https://commits.webkit.org/256191@main
